zepben/ewb-sdk-jvm
# EWB-4091 Tracing API v2

# Description

A complete rewrite of the Traversal / Tracing API. All traces have been replaced with a single `NetworkTrace` class that covers all the use cases for needing different traversal implementation in the old API while being easier to use and read. 

In the past you had to provide a custom `queueNext` function to the traversal if you just needed to tweak the logic of queuing items. This was very undesirable, but also meant reimplementing the logic for queuing connectivity. Getting connectivity also had multiple ways (per phase or ignore phase, terminal vs conducting equipment steps, tracking custom data in steps etc) and things are about to get a lot trickier when introducing Cut and Clamp into the model which doesn't use conventional connectivity node connectivity. This would have been a nightmare to manage and update all the custom queue next implementations, been extremely bug prone and likely would have made the tracing API too onerous to use properly.

The new `NetworkTrace` class addresses all of the above and while adding a bunch of quality-of-life things as well. It is network state aware (that is, it is easy to make it operate on either normal or current network state) and has a fluent API with a DSL style syntax. E.g. To create a downstream tree that stops after 5 equipment steps on phases AB you would go:

```kotlin
Tracing.networkTrace(networkStateOperators = NetworkStateOperators.CURRENT)
    .addCondition { stopAtOpen() } // will stop at currently open points due to the CURRENT state operators passed above
    .addCondition { downstream() } // likewise will trace currently downstream due to the state operators
    .addCondition { limitEquipmentSteps(5) }
    .addStepAction { step, context ->
        step.path.toTerminal // The terminal stepped to 
        step.path.toEquipment // The equipment stepped to 
        step.path.fromTerminal // The terminal stepped from (the previous step's toTerminal)
        step.path.fromEquipment // The equipment stepped from (the previous steps' toEquipment)
        step.path.phases // The phases traced on this equipment, empty if no specific phases were given in `run`)
        step.numEquipmentSteps // The number of equipment stepped on via the current trace path
        step.numTerminalSteps // The number of terminals stepped on via the current trace path
        step.type // INTERNAL if the step traversal across a piece of equipment, EXTERNAL if the step traversed to a new piece of euqipment
    }
    .run(terminal, phases = PhaseCode.AB)
```

This is extremely flexible compared to the old API where you the best you could do would be to use a `LimitedConnectedEquipmentTrace` where you could not add custom step actions nor specify phases. 

Likewise, you don't need different trace factories for upstream / downstream / normal / current etc. You just configure the NetworkTrace instance to behave how you need. The NetworkTrace instance configuration is very easy to read and see exactly what the conditions are without needing custom traversal implementations for different conditions.

This is obviously a massively breaking change, and all traces based on the old tracing API will need to be rewritten. However, porting internal code that uses the old tracing API has shown us that the new API leads to less lines of code overall that is easier to read and reason about.

# Associated tasks

There are plenty of downstream libraries that require updating for this change, however, they can list the PR as an associated task instead of trying to list all the effected libraries here (which don't all have PRs yet).

# Test Steps

The unit tests are pretty extensive as all existing tracing tests have been ported to use the new trace where appropriate. 
My suggestion would be to take a trace you have personally written with the old tracing API and porting it to use the `NetworkTrace` and check it behaves how you would expect. 

# Checklist

If any of these are not applicable, strikethrough the line `~like this~`. **Do not delete it!**. Let the reviewer decide if you should have done it.

### Code

### Documentation
- [ ] ~I have updated any documentation required for these changes.~ This is happening as a separate task I believe.

# Breaking Changes

Massively breaking change that will break anything using the tracing API in this SDK.
